src/aviary/functional.py

```diff
@@ -0,0 +1,173 @@
+from collections.abc import Callable
+from functools import update_wrapper
+from typing import Any
+
+from pydantic import BaseModel, Field
+
+from aviary.env import Environment, Frame
+from aviary.message import Message
+from aviary.tools import Messages, Tool, ToolRequestMessage
+from aviary.utils import is_coroutine_callable
+
+
+class DynamicState(BaseModel):
+    """Dynamic env state model that adapts to provided extras."""
+
+    reward: float = 0
+    done: bool = False
+    extras: dict[str, Any] = Field(default_factory=dict)
+
+    def __getattr__(self, name: str):
+        """Allow direct access to extras as attributes."""
+        if name in self.extras:
+            return self.extras[name]
+        raise AttributeError(f"'State' has no attribute {name!r}")
+
+
+class FunctionalEnvironment(Environment[DynamicState]):
+    """Environment class for function-based environments."""
+
+    def __init__(
+        self,
+        start_fn: Callable,
+        tools: list[Tool],
+        allow_concurrency: bool,
+        *args,
+        **kwargs,
+    ):
+        super().__init__()
+        self.start_fn = start_fn
+        self.tools = tools
+        self.args = args
+        self.kwargs = kwargs
+        self.state = DynamicState()
+        self.allow_concurrency = allow_concurrency
+
+    async def _call_start(self) -> tuple[str, dict[str, Any]]:
```

Comment:
```suggestion
    async def _call_start_fn(self) -> tuple[str, dict[str, Any]]:
```

Will help differentiate between `fenv.start`